src/components/transport_manager/src/transport_manager_impl.cc

```diff
@@ -413,14 +413,12 @@ int TransportManagerImpl::AddTransportAdapter(
                   "transport_adapter_listeners_.end()");
     return E_ADAPTER_EXISTS;
   }
-
-  if (transport_adapter->IsInitialised() ||
-      transport_adapter->Init() == TransportAdapter::OK) {
-    transport_adapter_listeners_[transport_adapter] =
+  transport_adapter_listeners_[transport_adapter] =
         new TransportAdapterListenerImpl(this, transport_adapter);
-    transport_adapter->AddListener(
+  transport_adapter->AddListener(
         transport_adapter_listeners_[transport_adapter]);
-
+  if (transport_adapter->IsInitialised() ||
+      transport_adapter->Init() == TransportAdapter::OK) {
     transport_adapters_.push_back(transport_adapter);
   } else {
```

Comment:
@BrandonHe is transport_adapter is not initialized shouldn't it be removed from ransport_adapter_listeners_? and shouldn't TransportAdapterListenerImpl be deleted? Please describe hot it fixes the bug

Reply:
@LuxoftAKutsan Here the condition is **or**, either transport_adapter is initialized or transport_adapter initialized status is OK, it will be added to transport_adapter_listeners. So if the transport_adapter is not initialized, it will run the transport_adapter->Init(). Otherwise, if neither transport_adapter is initialized nor transport_adapter initialization status is OK, it will not add to the transport_adapter_listeners. There is nothing need to be removed.

Yes, TransportAdapterListenerImpl  should not be deleted, it will help to create a new transport_adapter listeners, we didn't want to do that.

Here all the logic change is adding transport_adapter to transport_adapter_listeners before running transport_adapter->Init(), this will make transport_adapter be restored successfully.